src/kafka/protocol/kafka_primitives.hh

```diff
@@ -0,0 +1,321 @@
+
+namespace seastar {
+
+namespace kafka {
+
+struct parsing_exception : public std::exception {
+    [[nodiscard]] const char *what() const noexcept override {
+        return "Error parsing Kafka message.";
+    }
+};
+
+template<typename NumberType>
+class kafka_number_t {
+private:
+    NumberType _value;
+    static constexpr auto NUMBER_SIZE = sizeof(NumberType);
+
+public:
+    kafka_number_t() noexcept : kafka_number_t(0) {}
+
+    explicit kafka_number_t(NumberType value) noexcept : _value(value) {}
+
+    [[nodiscard]] const NumberType &operator*() const noexcept { return _value; }
+
+    [[nodiscard]] NumberType &operator*() noexcept { return _value; }
+
+    kafka_number_t &operator=(NumberType value) noexcept {
+        _value = value;
+        return *this;
+    }
+
+    void serialize(std::ostream &os, int16_t api_version) const {
+        std::array<char, NUMBER_SIZE> buffer{};
+        auto value = net::hton(_value);
+        auto value_pointer = reinterpret_cast<const char *>(&value);
+        std::copy(value_pointer, value_pointer + NUMBER_SIZE, buffer.begin());
+
+        os.write(buffer.data(), NUMBER_SIZE);
+    }
+
+    void deserialize(std::istream &is, int16_t api_version) {
+        std::array<char, NUMBER_SIZE> buffer{};
+        is.read(buffer.data(), NUMBER_SIZE);
+        if (is.gcount() != NUMBER_SIZE) {
+            throw parsing_exception();
+        }
+        _value = net::ntoh(*reinterpret_cast<NumberType *>(buffer.data()));
+    }
+};
+
+using kafka_int8_t = kafka_number_t<int8_t>;
+using kafka_int16_t = kafka_number_t<int16_t>;
+using kafka_int32_t = kafka_number_t<int32_t>;
+using kafka_int64_t = kafka_number_t<int64_t>;
+using kafka_uint32_t = kafka_number_t<uint32_t>;
+using kafka_bool_t = kafka_number_t<uint8_t>;
+
+template<typename SizeType>
+class kafka_buffer_t {
+private:
+    std::string _value;
+public:
+    kafka_buffer_t() noexcept = default;
+
+    explicit kafka_buffer_t(std::string value) : _value(std::move(value)) {}
+
+    [[nodiscard]] const std::string &operator*() const noexcept { return _value; }
+
+    [[nodiscard]] std::string &operator*() noexcept { return _value; }
+
+    [[nodiscard]] const std::string *operator->() const noexcept { return &_value; }
+
+    [[nodiscard]] std::string *operator->() noexcept { return &_value; }
+
+    kafka_buffer_t &operator=(const std::string &value) {
+        _value = value;
+        return *this;
+    }
+
+    kafka_buffer_t &operator=(std::string &&value) noexcept {
+        _value = std::move(value);
+        return *this;
+    }
+
+    void serialize(std::ostream &os, int16_t api_version) const {
+        kafka_number_t<SizeType> length(_value.size());
+        length.serialize(os, api_version);
+
+        os.write(_value.data(), _value.size());
+    }
+
+    void deserialize(std::istream &is, int16_t api_version) {
+        kafka_number_t<SizeType> length;
+        length.deserialize(is, api_version);
+        if (*length < 0) {
```

Comment:
does Kafka define max length? It would be better to check rather than accept a really large number here and crash the whole server with `std::bad_alloc`. You can also leave a TODO.

Reply:
It seems that there is no upper bound on length in protocol specification. Some messages (Fetch) allow you to specify max_bytes. Adding a TODO for now.